memwatch.1
.\" Manpage for memwatch
.TH MEMWATCH 1 "MAY 2016" Linux "User Manuals"
.SH NAME
memwatch \- Shows memory usage statistics in real time.
.SH SYNOPSIS
.B memwatch
.RI [ options ]
.SH DESCRIPTION
The main window displays the general statistics on the use of memory.
.TP
Data gather from /proc/meminfo.
.TP
The displayed keys for physical ram (Memory [unit size]):
.TP
\fBTotal\fR
Total usable ram.
.TP
\fBFree\fR
Unused memory.
.TP
\fBUsed\fR
Used memory (calculated as Total - Free - Buff - Cache).
.TP
\fBBuff\fR
Memory used by kernel buffers.
.TP
\fBCache\fR
Memory used by the page cache and slabs.
.TP
\fBAvail\fR
An estimate of how much memory is available for starting new applications, without swapping.
.TP
The displayed keys for swap (Swap [unit size]):
.TP
\fBTotal\fR
Total amount of swap space available.
.TP
\fBFree\fR
Unused swap space.
.TP
\fBUsed\fR
Used swap space.
.TP
Window of Process list prints processes, sorted by memory usage.
.TP
Data gather from /proc/PID/status and /proc/PID/cmdline.
.TP
\fBPID\fR
Process ID.
.TP
\fBOWNER\fR
Process owner.
.TP
\fBRES\fR
The size of use physical memory.
.TP
\fBSHR\fR
The size of the process's shared pages.
.TP
\fBVIRT\fR
The total amount of virtual memory used by the task.
.TP
\fBSWAP\fR
The size of the process's address space in swap.
.TP
\fBS\fR
Process status. Zombie processes are not displayed, since do not consume memory.
The status of the task which can be one of:
  \fBD\fR - uninterruptible sleep.
  \fBR\fR - running.
  \fBS\fR - sleeping.
  \fBT\fR - stopped by job control signal.
  \fBt\fR - stopped by debugger during trace.
.TP
\fBCommand\fR
The full command line of the process. Program name and arguments.
.TP
Process information window shows additional information.
.TP
Data gather from /proc/PID/status and /proc/PID/cmdline.
.TP
\fBVirtual, Peak\fR
Current virtual and peak vertual memory usage.
.TP
\fBResident, Peak\fR
Current physical and peak physical memory usage.
.TP
\fBShared\fR
The size of the process's shared pages.
.TP
\fBSwap\fR
The size of the process's address space in swap.
.TP
\fBData\fR
Size of data segments.
.TP
\fBStack\fR
Size of stack segments.
.TP
\fBText\fR
Size of text segments.
.TP
\fBLibrary\fR
Shared library code size.
.SH OPTIONS
.TP
\fB-d\fR, \fB\-\-delay\fR \fIseconds\fR
Set data update interval. Minimum value is 0.1, maximum is 60.
.TP
\fB-S\fR, \fB\-\-si\fR
Print sizes in powers of 1000 instead of 1024. In the International System of Units (SI) the prefix kilo means 1000.
.TP
\fB\-b\fR, \fB\-\-bytes\fR
Display the amount of memory in bytes.
.TP
\fB\-k\fR, \fB\-\-kilo\fR
Display the amount of memory in kilobytes.
.TP
\fB\-m\fR, \fB\-\-mega\fR
Display the amount of memory in megabytes.
.TP
\fB\-g\fR, \fB\-\-giga\fR
Display the amount of memory in gigabytes.
.TP
\fB\-t\fR, \fB\-\-tera\fR
Display the amount of memory in terabytes.
.TP
\fB\-h\fR, \fB\-\-human\fR
Display the amount of memory in human readable format.
.TP
\fB\-l\fR, \fB\-\-list\fR
Display a list of processes.
.TP
\fB-p\fR, \fB\-\-pid\fR \fIPID\fR
Show only processes with specified PID.
.TP
\fB\-V\fR, \fB\-\-version\fR
Prints version.
.TP
\fB\-\-help\fR
Prints help message.
.SH INTERACTIVE COMMANDS
.LP
The following commands are supported while in memwatch:
.LP
Global hotkeys:
.TP
.B b
Display the amount of memory in bytes.
.TP
.B k
Display the amount of memory in kilobytes.
.TP
.B m
Display the amount of memory in megabytes.
.TP
.B g
Display the amount of memory in gigabytes.
.TP
.B t
Display the amount of memory in terabytes.
.TP
.B h
Display the amount of memory in human readable format.
.TP
.B s
Toggle power size (1000 or 1024).
.TP
.B l
Switch between the main window and a list of processes.
.TP
.B F1, ?
Print help.
.TP
.B q
Exit memwatch.
.LP
Process list hotkeys:
.TP
.B Arrows
Scroll process list.
.TP
.B Home
Go to the begining of the process list.
.TP
.B End
Go to the end of the process list.
.TP
.B 1
Sort by RSS field.
.TP
.B 2
Sort by SHR field.
.TP
.B 3
Sort by VIRT field.
.TP
.B 4
Sort by SWAP field.
.TP
.B r
Switch the sort order.
.TP
.B Enter
Show information by selected process.
.TP
.B /
Look for a process by name.
.LP
Process information hotkeys:
.TP
.B Enter
Back to process list.
.SH AUTHOR
Pavel Balaev (pascal@unixdev.ru)


NEWS
v0.3.2 - 15 Aug 2016
--------------------
    - Look for a process by name.
    - Fixed distortion of the process table because of long usernames.
    - Internal ncurses Cmake module was removed.

v0.3.1 - 11 Jun 2016
--------------------
    - In the list of processes added hot keys: Home, End.
    - Man page was updated.

v0.3.0 - 27 May 2016
--------------------
    - Process list.
    - Show only processes with specified PID.
    - Help menu.
    - Fix bug with high cpu load.

v0.2.1 - 7 May 2016
--------------------
    - Fix bug with unit size options (bug #5).
    - React to the hot key is pressed immediately.
    - Man page was updated.

v0.2.0 - 1 May 2016
--------------------
    - Localization support (gettext).
    - Fix broken delay option.
    - Data update time is now set in seconds.
    - New field displays available memory.
    - Long options.
    - Size in powers of 1024 and 1000
    - Size in gigabytes and terabytes
    - Human readable output of memory size(enable by default)

v0.1.0 - 16 Apr 2016
--------------------
    - Use cmake instead of autotools.
    - Code refactoring.
    - Add hotkeys for display the amount of memory in Mb,Kb and bytes.

v0.0.2 - 03 Jun 2015
--------------------
    - Add Buffer and Cached memory info.

v0.0.1 - 19 May 2015
--------------------
    - initial commit


memwatch.c
/* This programm prints memory info in real time. 
 * compile: gcc memwatch.c -o memwatch -lcurses -lm -Wall 
 * author: PascalRD <pascal at unixdev ru> */

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <stdint.h>
#include <unistd.h>
#include <curses.h>
#include <string.h>
#include <math.h>
#include <libgen.h>
#include <errno.h>
#include <sys/time.h>

#include "memwatch.h"

int main(int argc, char **argv)
{
     int c;
     int opt, d_flag, u_delay;
     d_flag = 0;

     while((opt = getopt(argc, argv, "hvd:")) != -1)
     {
          switch(opt)
          {
               case 'd':
                    u_delay = atoi(optarg);
                    if(u_delay >= DELAY_MIN && u_delay <= DELAY_MAX)
                    {
                         d_flag = 1;
                    }
                    else
                    {
                         fprintf(stderr, "Supported delay range %d-%d\n", DELAY_MIN, DELAY_MAX);
                         exit(EXIT_FAILURE);
                    }
                    break;
               case 'v':
                         printf("memwatch v%s\n", VERSION);
                         exit(EXIT_SUCCESS);
               case 'h':
               default:
                    fprintf(stderr, "Usage: %s [-d ms]\n", basename(argv[0]));
                    exit(EXIT_FAILURE);
          }
     }

     d_flag ? set_timer(u_delay) : set_timer(DELAY);
     signal(SIGALRM, get_data);

     initscr();
     noecho();
     curs_set(0);
     cbreak();
     clear();

     get_data(0);

     while((c = getch()) != 'q')
     {
          pause();
     }

     endwin();

     return 0;
}

void set_timer(uint32_t usec)
{
     struct itimerval timer;

     uint32_t msec = (usec % 1000) * 1000;
     uint32_t sec  = usec / 1000;

     timer.it_interval.tv_sec = sec;
     timer.it_interval.tv_usec = msec;
     timer.it_value.tv_sec = sec;
     timer.it_value.tv_usec = msec;

     setitimer(ITIMER_REAL, &timer, NULL);
}

void get_data(UNUSED int signum)
{
     FILE *fp;
     char *line = NULL;
     size_t len = 0;
     ssize_t read;
     mem_t memory;

     memset(&memory, 0, sizeof(memory));
     memory.swap_disabled = 0;

     if((fp = fopen(MEMFILE, "r")) == NULL)
     {
          nocbreak();
          mvaddstr(1, 1, "Error opening file:");
          mvaddstr(2, 1, MEMFILE);
          getch();
          endwin();
          exit(EXIT_FAILURE);
     }

     while((read = getline(&line, &len, fp)) != -1)
     {
          if(strncmp("MemFree:", line, strlen("MemFree:")) == 0)
          {
               insert_value(&memory, line, FREE_MEM);
          }
          else if(strncmp("MemTotal:", line, strlen("MemTotal:")) == 0)
          {
               insert_value(&memory, line, TOTAL_MEM);
          }
          else if(strncmp("Buffers:", line, strlen("Buffers:")) == 0)
          {
               insert_value(&memory, line, BUFF_MEM);
          }
          else if(strncmp("Cached:", line, strlen("Cached:")) == 0)
          {
               insert_value(&memory, line, CACHE_MEM);
          }
          else if(strncmp("SwapFree:", line, strlen("SwapFree:")) == 0)
          {
               insert_value(&memory, line, FREE_SWAP);
          }
          else if(strncmp("SwapTotal:", line, strlen("SwapTotal:")) == 0)
          {
               insert_value(&memory, line, TOTAL_SWAP);
          }
     }

     free(line);
     fclose(fp);

     print_info(&memory);
}

void print_info(const mem_t *mem)
{
     char buf[BUFDEC] = {0};
     float mem_ratio = (float) mem->mem_used / mem->mem_total;
     float mem_bar_used = BAR_LEHGTH * mem_ratio;
     uint32_t mem_bar_free = BAR_LEHGTH - floor(mem_bar_used);
     float swap_ratio = (float) mem->swap_used / mem->swap_total;
     float swap_bar_used = BAR_LEHGTH * swap_ratio;
     uint32_t swap_bar_free = BAR_LEHGTH - floor(swap_bar_used);
     size_t i;
     uint16_t col = 1;

     for(i = 1; i <= 20; i++)
     {
          mvaddstr(i, 1, "                                        ");
     }

     mvaddstr(col, 1, "Memory:");
     col++; col++;
     print_bar(col, mem_bar_used, mem_bar_free);
     col++; col++;
     mvaddstr(col, 1, "Total");
     mvaddstr(col++, 10, num_to_str(buf, mem->mem_total));
     mvaddstr(col, 1, "Free");
     mvaddstr(col++, 10, num_to_str(buf, mem->mem_free));
     mvaddstr(col, 1, "Used");
     mvaddstr(col++, 10, num_to_str(buf, mem->mem_used));
     mvaddstr(col, 1, "Buff");
     mvaddstr(col++, 10, num_to_str(buf, mem->mem_buff));
     mvaddstr(col, 1, "Cache");
     mvaddstr(col++, 10, num_to_str(buf, mem->mem_cache));

     if(mem->swap_disabled)
     {
          col++; col++;
          mvaddstr(col, 1, "Swap disabled.");
     }
     else
     {
          col++; col++;
          mvaddstr(col, 1, "Swap:");
          col++; col++;
          print_bar(col, swap_bar_used, swap_bar_free);
          col++; col++;
          mvaddstr(col, 1, "Total");
          mvaddstr(col++, 10, num_to_str(buf, mem->swap_total));
          mvaddstr(col, 1, "Free");
          mvaddstr(col++, 10, num_to_str(buf, mem->swap_free));
          mvaddstr(col, 1, "Used");
          mvaddstr(col, 10, num_to_str(buf, mem->swap_used));
     }

     refresh();
}

void print_bar(uint32_t col, uint32_t used, uint32_t last)
{
     size_t i, row;
     row = 2;

     mvaddch(col, row - 1, '[');
     for(i = 0; i < used; i++, row++)
     {
          mvaddch(col, row, '#');
     }
     for(i = 0; i < last; i++, row++)
     {
          mvaddch(col, row, '-');
     }
     mvaddch(col, row, ']');
}

void insert_value(mem_t *mem, char *line, int ch)
{
     char *token;
     char *end = NULL;
     char buf[BUFDEC];
     uint64_t num;
     errno = 0;

     token = strtok(line, " ");
     token = strtok(NULL, " ");
     num = strtoull(token, &end, 10);
     if((token + strlen(token)) != end || errno == ERANGE)
     {
          num = 0;
     }
     else
     {
          num /= DELIM;
     }
     snprintf(buf, sizeof(buf), "%lu", num);

     switch(ch)
     {
          case FREE_MEM:
               mem->mem_free = num;
               mem->mem_used = mem->mem_total - mem->mem_free;
               break;
          case FREE_SWAP:
               if(! mem->swap_disabled)
               {
                    mem->swap_free = num;
                    mem->swap_used = mem->swap_total - mem->swap_free;
               }
               break;
          case TOTAL_MEM:
               mem->mem_total = num;
               break;
          case BUFF_MEM:
               mem->mem_buff = num;
               break;
          case CACHE_MEM:
               mem->mem_cache = num;
               break;
          case TOTAL_SWAP:
               if(! num)
               {
                    mem->swap_disabled = 1;
               }
               else
               {
                    mem->swap_total = num;
               }
               break;
     }
}

char *num_to_str(char* buf, uint64_t num)
{
    snprintf(buf, BUFDEC, "%lu", num);
    return buf;
}
